### ISSUE TITLE
The shape of AiState

### ISSUE THREAD

**Issue by mlutze** [mlutze/flixball#57]
`AiState`'s definition is this:

```scala
pub type alias AiState = (Int32, Map[String, String])
```
where the `Int32` part is for easy counting, etc. and the `Map` part is for arbitrary storage.

It's kind of weird to have a "int part" and "everything else part".

I propose instead the following:
```scala
pub enum AiState {
    case Value(String)
    case Collection(List[AiState])
}
```

This makes it a simple enum and conversions to most types come quite naturally. I was originally considering a JSON style but all the types add unnecessary complexity. We lose some guarantees in the conversion for integers, due to malformed encoding, but that was always going to be a potential issue with anything nontrivial anyway, so I don't have a problem with just falling back to some default value in this case.

Serialization of this is also very easy, as we can just do, e.g. a Lisp Syntax.

**Comment by mlutze** [comment]
If AiState is just String, then all the conversion functions `AiState -> CustomState` will have to parse the string, which gets annoying with more complex types. With this intermediary, we just have one function that has to do string parsing (for reading state from a file).

**Comment by JonathanStarup** [comment]
> If AiState is just String, then all the conversion functions `AiState -> CustomState` will have to parse the string, which gets annoying with more complex types. With this intermediary, we just have one function that has to do string parsing (for reading state from a file).

i see the point, but this half-way conversion can still exist without it being stored on the state. just having a string state seems less arbitrary

**Comment by mlutze** [comment]
I've given this some thought. I agree that a string would be a clean solution, but I'm worried that it allows for too much freedom, which might hurt us when dealing with more complex types. Any parsing function requires all the sub-parsing-functions that it calls to be defined via the same "model", otherwise it all breaks.

As an extreme example, say I have my custom representation:
```scala

enum Blah

instance Stateable[Blah] {
    def toState(x: Blah): AiState = "\"hello\", [], ][, [][],"
}
```

and some representation for a list
```scala
instance Stateable[List[a]] with Stateable[a] {
    def toState(x: List[a]): AiState = "[" + String.intercalate(", ", List.map(toState, x) + "]"
}
```

The result of `toState(Blah :: Nil)` is very hard to parse: `"[\"hello\", [], ]][, [][],]"`

On the other hand, if we enforce more of a structure:
```scala
instance Stateable[Blah] {
    def toState(x: Blah): AiState = Value("\"hello\", [], ][, [][],")
}

instance Stateable[List[a]] with Stateable[a] {
    def toState(x: List[a]): AiState = Collection(List.map(toState, x))
}
```

then we know what to expect when parsing.

**Comment by JonathanStarup** [comment]
But the parsing responsibility is on the types themselves right? `Blah` also has to implement
```
instance UnStateable[Blah] {
    def fromState(as: AiState): Blah = ...
}

```
